[Cases] [Observability] Disable add-to-case functionality when all selected alerts are already attached

justinkambic opened the pull request:
## Summary

Resolves https://github.com/elastic/kibana/issues/227036.

For o11y alerting, we want to make it so that if a user selects a set of cases, and the full set of selected alerts is already on the case, the add to case button will be disabled on the add-to-existing-case modal.

~This functionality is transparent to other consumers of this feature~ This functionality is driven by the `getAttachments` function implemented by the caller of the _Add to existing case_ modal. The expectation is that `getAttachments` will provide a list of all the selected attachments.

This code then queries all of the existing user comments of type `alert` attached to the cases appearing in the modal. The code diffs the selected alerts against the existing attachments. If all the selected alerts intersect with the existing case attachments, that case's button will be disabled and include copy and an icon to indicate that it already contains the selected alerts.

**Caveat:** This feature as it is written only works for cases with 100 or fewer attached alerts. In the case where a case has > 100 alerts attached to it, the functionality will work as it does today and the user can click the Add button. The code already filters any existing cases further down the procedure, and I am assuming most cases have fewer than 100 alerts attached in today's common usage, so this tradeoff seemed preferable to adding the complexity and querying load of pagination logic.

<img width="1165" height="391" alt="image" src="https://github.com/user-attachments/assets/bf0b5daa-4b3c-452b-889b-ccd4a2f9906d" />

justinkambic committed:
Add filtering to disable add-to-case functionality for observability alerts where all selected alerts were previously added.

elasticmachine commented:
Pinging @elastic/obs-ux-management-team (Team:obs-ux-management)

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/329837)
* Commit: fc9be6f970af362920f1db6f23947422e83a2645

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Module Count
> Fewer modules leads to a faster build time


| id | [before](https://github.com/elastic/kibana/commit/f120122938eeae723bd76b1d0601d2baf323d342) | [after](https://github.com/elastic/kibana/commit/fc9be6f970af362920f1db6f23947422e83a2645) | diff |
| --- | --- | --- | --- |
| `cases` | 1131 | 1132 | +1 |

#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/f120122938eeae723bd76b1d0601d2baf323d342) | [after](https://github.com/elastic/kibana/commit/fc9be6f970af362920f1db6f23947422e83a2645) | diff |
| --- | --- | --- | --- |
| `cases` | 1.3MB | 1.3MB | +438.0B |

#### Public APIs missing exports
> Total count of every type that is part of your API that should be exported but is not. This will cause broken links in the API documentation system. Target amount is 0. `Run node scripts/build_api_docs --plugin [yourplugin] --stats exports` for more detailed information.


| id | [before](https://github.com/elastic/kibana/commit/f120122938eeae723bd76b1d0601d2baf323d342) | [after](https://github.com/elastic/kibana/commit/fc9be6f970af362920f1db6f23947422e83a2645) | diff |
| --- | --- | --- | --- |
| `cases` | 28 | 29 | +1 |

#### Page load bundle
> Size of the bundles that are downloaded on every page load. Target size is below 100kb


| id | [before](https://github.com/elastic/kibana/commit/f120122938eeae723bd76b1d0601d2baf323d342) | [after](https://github.com/elastic/kibana/commit/fc9be6f970af362920f1db6f23947422e83a2645) | diff |
| --- | --- | --- | --- |
| `cases` | 135.2KB | 135.4KB | +145.0B |



cc @justinkambic

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/329837","number":329837,"commit":"fc9be6f970af362920f1db6f23947422e83a2645"}],"number":329837}
buildkite-pr-comment-->

janmonschke commented:
@justinkambic I desk-tested this but it did not work for me. Is there sth that I'm missing?


https://github.com/user-attachments/assets/af9b2d9d-b023-4b21-b19e-62ea7a1493d2

justinkambic committed:
Update check alert attachments hook.

Fix issues and clean up.

Introduce a loading flag for the check attachments hook.

Move loading flag to correct prop.

Refactor useCheckAlertAttachments for improved clarity and maintainability

justinkambic commented:
@janmonschke thanks for the call-out, there were some false assumptions I made. I've updated this to query the attachments per case. LMK if you're still seeing issues in practical testing.

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/331268)
* Commit: 614b5414a3d42eb426347e3873444ad079b54732

### Failed CI Steps
* [FTR Configs #53](https://buildkite.com/elastic/kibana-pull-request/builds/331268#0198c876-893d-4330-b8e2-08cb74b47342)
* [FTR Configs #53](https://buildkite.com/elastic/kibana-pull-request/builds/331268#0198c89c-cd71-4e35-a77a-71e91ca40125)
* [FTR Configs #76](https://buildkite.com/elastic/kibana-pull-request/builds/331268#0198c876-8957-4d5c-82dd-84d00394c712)

### Test Failures
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/331268#0198c876-893d-4330-b8e2-08cb74b47342) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/331268/jobs/0198c876-893d-4330-b8e2-08cb74b47342/artifacts/0198c89c-adc4-48da-93cf-6b2939babebd) FTR Configs #<span></span>53 / lens serverless - group 1 - subgroup 5 lens tsdb downsampling "after all" hook in "downsampling"
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/331268#0198c876-8957-4d5c-82dd-84d00394c712) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/331268/jobs/0198c876-8957-4d5c-82dd-84d00394c712/artifacts/0198c89b-58b3-49d2-b602-d2b016d6ce21) FTR Configs #<span></span>76 / lens serverless - group 1 - subgroup 5 lens tsdb downsampling "after all" hook in "downsampling"
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/331268#0198c89c-cd71-4e35-a77a-71e91ca40125) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/331268/jobs/0198c89c-cd71-4e35-a77a-71e91ca40125/artifacts/0198c8a9-0a71-4043-8140-174f6e7ab2c8) FTR Configs #<span></span>53 / lens serverless - group 1 - subgroup 5 lens tsdb downsampling "after all" hook in "downsampling"
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/331268#0198c876-893d-4330-b8e2-08cb74b47342) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/331268/jobs/0198c876-893d-4330-b8e2-08cb74b47342/artifacts/0198c89c-adc5-42b7-8ed4-a698a7b22079) FTR Configs #<span></span>53 / lens serverless - group 1 - subgroup 5 lens tsdb downsampling "before all" hook in "downsampling"
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/331268#0198c876-8957-4d5c-82dd-84d00394c712) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/331268/jobs/0198c876-8957-4d5c-82dd-84d00394c712/artifacts/0198c89b-58b3-4fd6-aeb5-c92046952a25) FTR Configs #<span></span>76 / lens serverless - group 1 - subgroup 5 lens tsdb downsampling "before all" hook in "downsampling"
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/331268#0198c89c-cd71-4e35-a77a-71e91ca40125) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/331268/jobs/0198c89c-cd71-4e35-a77a-71e91ca40125/artifacts/0198c8a9-0a71-4380-b84b-3be2215d6f54) FTR Configs #<span></span>53 / lens serverless - group 1 - subgroup 5 lens tsdb downsampling "before all" hook in "downsampling"

### History
* :broken_heart: [Build #330951](https://buildkite.com/elastic/kibana-pull-request/builds/330951) failed 22b76d9508eff97fbbf560fe840182d6969e8aa5
* :green_heart: [Build #329837](https://buildkite.com/elastic/kibana-pull-request/builds/329837) succeeded fc9be6f970af362920f1db6f23947422e83a2645

cc @justinkambic

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/331268","number":331268,"commit":"614b5414a3d42eb426347e3873444ad079b54732"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/330951","number":330951,"commit":"22b76d9508eff97fbbf560fe840182d6969e8aa5"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/329837","number":329837,"commit":"fc9be6f970af362920f1db6f23947422e83a2645"}],"number":331268}
buildkite-pr-comment-->

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/332335)
* Commit: eef6c14aff98c25d61a59f10141e97930b837fa4

### Failed CI Steps
* [Check Types](https://buildkite.com/elastic/kibana-pull-request/builds/332335#0198e214-2066-4442-8ea2-3a344dd86da0)

### History
* :yellow_heart: [Build #331445](https://buildkite.com/elastic/kibana-pull-request/builds/331445) was flaky 7fa2af2a2721e41d0a2865b2cfe3a7073ebdfe52
* :broken_heart: [Build #331268](https://buildkite.com/elastic/kibana-pull-request/builds/331268) failed 614b5414a3d42eb426347e3873444ad079b54732
* :broken_heart: [Build #330951](https://buildkite.com/elastic/kibana-pull-request/builds/330951) failed 22b76d9508eff97fbbf560fe840182d6969e8aa5
* :green_heart: [Build #329837](https://buildkite.com/elastic/kibana-pull-request/builds/329837) succeeded fc9be6f970af362920f1db6f23947422e83a2645

cc @justinkambic

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/332335","number":332335,"commit":"eef6c14aff98c25d61a59f10141e97930b837fa4"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/331445","number":331445,"commit":"7fa2af2a2721e41d0a2865b2cfe3a7073ebdfe52"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/331268","number":331268,"commit":"614b5414a3d42eb426347e3873444ad079b54732"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/330951","number":330951,"commit":"22b76d9508eff97fbbf560fe840182d6969e8aa5"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/329837","number":329837,"commit":"fc9be6f970af362920f1db6f23947422e83a2645"}],"number":332335}
buildkite-pr-comment-->

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/332342)
* Commit: c8a59e8a227139416a3c4b30b2d44008c0fa4e15

### Failed CI Steps
* [Check Types](https://buildkite.com/elastic/kibana-pull-request/builds/332342#0198e22a-79fd-4c02-ac07-1c4d9c50dbe3)

### History
* :broken_heart: [Build #332335](https://buildkite.com/elastic/kibana-pull-request/builds/332335) failed eef6c14aff98c25d61a59f10141e97930b837fa4
* :yellow_heart: [Build #331445](https://buildkite.com/elastic/kibana-pull-request/builds/331445) was flaky 7fa2af2a2721e41d0a2865b2cfe3a7073ebdfe52
* :broken_heart: [Build #331268](https://buildkite.com/elastic/kibana-pull-request/builds/331268) failed 614b5414a3d42eb426347e3873444ad079b54732
* :broken_heart: [Build #330951](https://buildkite.com/elastic/kibana-pull-request/builds/330951) failed 22b76d9508eff97fbbf560fe840182d6969e8aa5
* :green_heart: [Build #329837](https://buildkite.com/elastic/kibana-pull-request/builds/329837) succeeded fc9be6f970af362920f1db6f23947422e83a2645

cc @justinkambic

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/332342","number":332342,"commit":"c8a59e8a227139416a3c4b30b2d44008c0fa4e15"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/332335","number":332335,"commit":"eef6c14aff98c25d61a59f10141e97930b837fa4"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/331445","number":331445,"commit":"7fa2af2a2721e41d0a2865b2cfe3a7073ebdfe52"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/331268","number":331268,"commit":"614b5414a3d42eb426347e3873444ad079b54732"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/330951","number":330951,"commit":"22b76d9508eff97fbbf560fe840182d6969e8aa5"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/329837","number":329837,"commit":"fc9be6f970af362920f1db6f23947422e83a2645"}],"number":332342}
buildkite-pr-comment-->

justinkambic committed:
Add API route and logic to find cases by attachment IDs.

justinkambic committed:
Refactor useCheckAlertAttachments to utilize useFindCasesContainingAllSelectedAlerts and remove useFindCasesUserActions

justinkambic committed:
Add findCasesContainingAllAlerts route and associated tests

justinkambic committed:
Update INTERNAL_CASE_GET_CASES_BY_ATTACHMENT_URL to find cases by alerts

justinkambic committed:
Refactor findCasesByAttachmentId to use POST method and update request handling for alertIds and caseIds

justinkambic committed:
Fix test.

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/332886)
* Commit: f19981d354d5038f3a6a4b772b0011140fd1e789

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Module Count
> Fewer modules leads to a faster build time


| id | [before](https://github.com/elastic/kibana/commit/80d4e452cec93e3d45d155affaa6ae4b3b31d282) | [after](https://github.com/elastic/kibana/commit/f19981d354d5038f3a6a4b772b0011140fd1e789) | diff |
| --- | --- | --- | --- |
| `cases` | 1132 | 1134 | +2 |

#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/80d4e452cec93e3d45d155affaa6ae4b3b31d282) | [after](https://github.com/elastic/kibana/commit/f19981d354d5038f3a6a4b772b0011140fd1e789) | diff |
| --- | --- | --- | --- |
| `cases` | 1.3MB | 1.3MB | +900.0B |

#### Public APIs missing exports
> Total count of every type that is part of your API that should be exported but is not. This will cause broken links in the API documentation system. Target amount is 0. `Run node scripts/build_api_docs --plugin [yourplugin] --stats exports` for more detailed information.


| id | [before](https://github.com/elastic/kibana/commit/80d4e452cec93e3d45d155affaa6ae4b3b31d282) | [after](https://github.com/elastic/kibana/commit/f19981d354d5038f3a6a4b772b0011140fd1e789) | diff |
| --- | --- | --- | --- |
| `cases` | 28 | 29 | +1 |

#### Page load bundle
> Size of the bundles that are downloaded on every page load. Target size is below 100kb


| id | [before](https://github.com/elastic/kibana/commit/80d4e452cec93e3d45d155affaa6ae4b3b31d282) | [after](https://github.com/elastic/kibana/commit/f19981d354d5038f3a6a4b772b0011140fd1e789) | diff |
| --- | --- | --- | --- |
| `cases` | 135.3KB | 135.8KB | +429.0B |



### History
* :broken_heart: [Build #332342](https://buildkite.com/elastic/kibana-pull-request/builds/332342) failed c8a59e8a227139416a3c4b30b2d44008c0fa4e15
* :broken_heart: [Build #332335](https://buildkite.com/elastic/kibana-pull-request/builds/332335) failed eef6c14aff98c25d61a59f10141e97930b837fa4
* :yellow_heart: [Build #331445](https://buildkite.com/elastic/kibana-pull-request/builds/331445) was flaky 7fa2af2a2721e41d0a2865b2cfe3a7073ebdfe52
* :broken_heart: [Build #331268](https://buildkite.com/elastic/kibana-pull-request/builds/331268) failed 614b5414a3d42eb426347e3873444ad079b54732
* :broken_heart: [Build #330951](https://buildkite.com/elastic/kibana-pull-request/builds/330951) failed 22b76d9508eff97fbbf560fe840182d6969e8aa5

cc @justinkambic

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/332886","number":332886,"commit":"f19981d354d5038f3a6a4b772b0011140fd1e789"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/332342","number":332342,"commit":"c8a59e8a227139416a3c4b30b2d44008c0fa4e15"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/332335","number":332335,"commit":"eef6c14aff98c25d61a59f10141e97930b837fa4"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/331445","number":331445,"commit":"7fa2af2a2721e41d0a2865b2cfe3a7073ebdfe52"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/331268","number":331268,"commit":"614b5414a3d42eb426347e3873444ad079b54732"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/330951","number":330951,"commit":"22b76d9508eff97fbbf560fe840182d6969e8aa5"}],"number":332886}
buildkite-pr-comment-->

justinkambic committed:
Refactor findCasesByAttachmentId and useFindCasesContainingAllSelectedAlerts to improve type definitions for query results

justinkambic commented on x-pack/platform/plugins/shared/cases/public/containers/use_find_cases_containing_all_selected_alerts.tsx:
```diff
@@ -0,0 +1,24 @@
+  selectedAlertIds: string[],
+  caseIds: string[]
+) => {
+  return useQuery<unknown, unknown, { casesWithAllAttachments?: string[] }>({
```
[2423313](https://github.com/elastic/kibana/pull/231877/commits/2423313b4ebd23c96025c6d1b88f71af2dc2a614)

justinkambic commented on x-pack/platform/plugins/shared/cases/public/containers/api.ts:
```diff
@@ -221,6 +222,16 @@ export const getInferenceConnectors = async (
 };
 
+export const findCasesByAttachmentId = async (alertIds: string[], caseIds: string[]) => {
+  return KibanaServices.get().http.fetch<{}>(`${INTERNAL_CASE_GET_CASES_BY_ATTACHMENT_URL}`, {
```
[2423313](https://github.com/elastic/kibana/pull/231877/commits/2423313b4ebd23c96025c6d1b88f71af2dc2a614)

janmonschke commented on x-pack/platform/plugins/shared/cases/server/routes/api/internal/find_cases_containing_all_alerts.ts:
```diff
@@ -0,0 +1,108 @@
+) => {
+  const caseAlertIds = new Set<string>();
+
+  const alertsForCase = await casesClient.attachments.getAllAlertsAttachToCase({
+    caseId,
+  });
+
+  // there are more selected alerts than attached alerts
+  if (alertIds.size > alertsForCase.length) return null;
+
+  // we must walk the case's attached alerts
+  for (const alert of alertsForCase) {
+    if (alertIds.has(alert.id)) {
+      caseAlertIds.add(alert.id);
+    }
+  }
```
Instead of traversing the alert attachments manually, it would be great if the Elasticsearch query could do the heavy lifting for us.

One way to achieve that is by adding a filter to add an optional filter array to `getAllAlertsAttachToCase` here:
https://github.com/elastic/kibana/blob/7acaf614239637ccde03a0d5049a4c5062427ca6/x-pack/platform/plugins/shared/cases/server/client/attachments/get.ts#L65-L69

That filter then needs to be passed to `getAllAlertsAttachToCase` in the `AttachmentGetter` here:

https://github.com/elastic/kibana/blob/7acaf614239637ccde03a0d5049a4c5062427ca6/x-pack/platform/plugins/shared/cases/server/services/attachments/operations/get.ts#L141-L144

The filter array can be built like the alert type filter here (only that one filter is needed per alert id):
https://github.com/elastic/kibana/blob/7acaf614239637ccde03a0d5049a4c5062427ca6/x-pack/platform/plugins/shared/cases/server/services/attachments/operations/get.ts#L147-L152

The purpose of this filter is to only include documents that have one of the alert id's as their `alertId` prop. This will reduce the worst case Kibana server runtime for this query by a lot and it will in turn speed up the Elasticsearch query because not all alert attachments will be returned.

justinkambic replied:
Thanks this was really good feedback.

This also means we can trust the Elasticsearch filtering and do a `total` comparison between the selected alert IDs and the length of the result set, and skip any comparison in Kibana code altogether.

Will push soon.

Example output:

### One alert selected

<img width="472" height="173" alt="image" src="https://github.com/user-attachments/assets/f05079a9-dd67-48b9-8423-1963e55266e2" />

### Many alerts selected

<img width="465" height="404" alt="image" src="https://github.com/user-attachments/assets/616e1b13-aa5d-4c4d-acb9-22a9a85a9743" />

<img width="469" height="834" alt="image" src="https://github.com/user-attachments/assets/e47877f6-5bdd-4710-834f-9e1bdfe3e044" />

justinkambic replied:
[c5dd201](https://github.com/elastic/kibana/pull/231877/commits/c5dd201d27150484052aaef1bd05da92ed65c267)

janmonschke commented on x-pack/platform/plugins/shared/cases/server/routes/api/internal/find_cases_containing_all_alerts.ts:
```diff
@@ -0,0 +1,108 @@
+  security: DEFAULT_CASES_ROUTE_SECURITY,
+  params: {
+    body: schema.object({
+      alertIds: schema.maybe(schema.oneOf([schema.string(), schema.arrayOf(schema.string())])),
+      caseIds: schema.maybe(schema.oneOf([schema.string(), schema.arrayOf(schema.string())])),
+    }),
```
Is there a scenario where either `alertIds` or `caseIds` is not defined or if they're strings?
I'm asking because we can make the body parser quite strict and then won't need to add defensive code further down in the route handler.

For example this one can be removed then because the param parser takes care of that work for us:

```ts
if (!isStringOrArray(caseIds) || !isStringOrArray(alertIds)) {
      return response.badRequest({
        body: { message: 'Invalid request parameters' },
      });
    }
```

elasticmachine commented:
## :yellow_heart: Build succeeded, but was flaky
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/333263)
* Commit: cc16cab038e76d29ed018a41b4f52361d743ebb7

### Failed CI Steps
* [FTR Configs #7](https://buildkite.com/elastic/kibana-pull-request/builds/333263#0198ec9f-f726-465a-a543-fc5732a7630f)

### Test Failures
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/333263#0198ec9f-f726-465a-a543-fc5732a7630f) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/333263/jobs/0198ec9f-f726-465a-a543-fc5732a7630f/artifacts/0198ecd5-3e65-43d8-b023-22c4197469ca) FTR Configs #<span></span>7 / Search solution tests Search Home page Solution Nav - Search search home page with existing indices "before each" hook for "displays the feedbacks link"

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Module Count
> Fewer modules leads to a faster build time


| id | [before](https://github.com/elastic/kibana/commit/d02dd20b4c0724d21cda33b69c4fc926f48ae827) | [after](https://github.com/elastic/kibana/commit/cc16cab038e76d29ed018a41b4f52361d743ebb7) | diff |
| --- | --- | --- | --- |
| `cases` | 1152 | 1154 | +2 |

#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/d02dd20b4c0724d21cda33b69c4fc926f48ae827) | [after](https://github.com/elastic/kibana/commit/cc16cab038e76d29ed018a41b4f52361d743ebb7) | diff |
| --- | --- | --- | --- |
| `cases` | 1.4MB | 1.4MB | +875.0B |

#### Public APIs missing exports
> Total count of every type that is part of your API that should be exported but is not. This will cause broken links in the API documentation system. Target amount is 0. `Run node scripts/build_api_docs --plugin [yourplugin] --stats exports` for more detailed information.


| id | [before](https://github.com/elastic/kibana/commit/d02dd20b4c0724d21cda33b69c4fc926f48ae827) | [after](https://github.com/elastic/kibana/commit/cc16cab038e76d29ed018a41b4f52361d743ebb7) | diff |
| --- | --- | --- | --- |
| `cases` | 28 | 29 | +1 |

#### Page load bundle
> Size of the bundles that are downloaded on every page load. Target size is below 100kb


| id | [before](https://github.com/elastic/kibana/commit/d02dd20b4c0724d21cda33b69c4fc926f48ae827) | [after](https://github.com/elastic/kibana/commit/cc16cab038e76d29ed018a41b4f52361d743ebb7) | diff |
| --- | --- | --- | --- |
| `cases` | 137.5KB | 138.0KB | +429.0B |



### History
* :green_heart: [Build #332886](https://buildkite.com/elastic/kibana-pull-request/builds/332886) succeeded f19981d354d5038f3a6a4b772b0011140fd1e789
* :broken_heart: [Build #332342](https://buildkite.com/elastic/kibana-pull-request/builds/332342) failed c8a59e8a227139416a3c4b30b2d44008c0fa4e15
* :broken_heart: [Build #332335](https://buildkite.com/elastic/kibana-pull-request/builds/332335) failed eef6c14aff98c25d61a59f10141e97930b837fa4
* :yellow_heart: [Build #331445](https://buildkite.com/elastic/kibana-pull-request/builds/331445) was flaky 7fa2af2a2721e41d0a2865b2cfe3a7073ebdfe52
* :broken_heart: [Build #331268](https://buildkite.com/elastic/kibana-pull-request/builds/331268) failed 614b5414a3d42eb426347e3873444ad079b54732
* :broken_heart: [Build #330951](https://buildkite.com/elastic/kibana-pull-request/builds/330951) failed 22b76d9508eff97fbbf560fe840182d6969e8aa5

cc @justinkambic

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/333263","number":333263,"commit":"cc16cab038e76d29ed018a41b4f52361d743ebb7"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/332886","number":332886,"commit":"f19981d354d5038f3a6a4b772b0011140fd1e789"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/332342","number":332342,"commit":"c8a59e8a227139416a3c4b30b2d44008c0fa4e15"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/332335","number":332335,"commit":"eef6c14aff98c25d61a59f10141e97930b837fa4"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/331445","number":331445,"commit":"7fa2af2a2721e41d0a2865b2cfe3a7073ebdfe52"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/331268","number":331268,"commit":"614b5414a3d42eb426347e3873444ad079b54732"}],"number":333263}
buildkite-pr-comment-->

justinkambic committed:
Stricter endpoint validation.

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/333750)
* Commit: 68410af1e491a6b8d92aafac3c5dabffbc8c50da

### Failed CI Steps
* [Quick Checks](https://buildkite.com/elastic/kibana-pull-request/builds/333750#0198f165-d23d-459f-80a8-a77ec8ede2ae)

### History
* :yellow_heart: [Build #333263](https://buildkite.com/elastic/kibana-pull-request/builds/333263) was flaky cc16cab038e76d29ed018a41b4f52361d743ebb7
* :green_heart: [Build #332886](https://buildkite.com/elastic/kibana-pull-request/builds/332886) succeeded f19981d354d5038f3a6a4b772b0011140fd1e789
* :broken_heart: [Build #332342](https://buildkite.com/elastic/kibana-pull-request/builds/332342) failed c8a59e8a227139416a3c4b30b2d44008c0fa4e15
* :broken_heart: [Build #332335](https://buildkite.com/elastic/kibana-pull-request/builds/332335) failed eef6c14aff98c25d61a59f10141e97930b837fa4
* :yellow_heart: [Build #331445](https://buildkite.com/elastic/kibana-pull-request/builds/331445) was flaky 7fa2af2a2721e41d0a2865b2cfe3a7073ebdfe52
* :broken_heart: [Build #331268](https://buildkite.com/elastic/kibana-pull-request/builds/331268) failed 614b5414a3d42eb426347e3873444ad079b54732

cc @justinkambic

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/333750","number":333750,"commit":"68410af1e491a6b8d92aafac3c5dabffbc8c50da"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/333263","number":333263,"commit":"cc16cab038e76d29ed018a41b4f52361d743ebb7"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/332886","number":332886,"commit":"f19981d354d5038f3a6a4b772b0011140fd1e789"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/332342","number":332342,"commit":"c8a59e8a227139416a3c4b30b2d44008c0fa4e15"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/332335","number":332335,"commit":"eef6c14aff98c25d61a59f10141e97930b837fa4"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/331445","number":331445,"commit":"7fa2af2a2721e41d0a2865b2cfe3a7073ebdfe52"}],"number":333750}
buildkite-pr-comment-->

justinkambic committed:
Enhance getAllAlertsAttachToCase to support optional filter parameter and update processCase to utilize it for improved alert retrieval.

justinkambic committed:
Refactor tests for findCasesContainingAllAlerts to simplify alert retrieval logic and improve clarity in assertions.

janmonschke commented on x-pack/platform/plugins/shared/cases/public/containers/use_check_alert_attachments.tsx:
```diff
@@ -0,0 +1,44 @@
+    cases.map(({ id }) => id)
+  );
+
+  return { disabledCases: new Set(data?.casesWithAllAttachments ?? []), isLoading: isFetching };
```
Could you memoize the `new Set` part?

janmonschke reviewed (approved):
This is such a nice quality of life improvement 🎉 👏 

I left a couple small comments that should ideally be fixed before merging. It would be great to get one of your colleagues to review the PR as well.

justinkambic committed:
Memoize set creation in `useCheckAlertAttachments`.

justinkambic committed:
Include description for new translation.

justinkambic commented on x-pack/platform/plugins/shared/cases/public/components/all_cases/translations.ts:
```diff
@@ -96,6 +96,10 @@ export const SELECT = i18n.translate('xpack.cases.caseTable.select', {
 });
 
+export const ALREADY_ATTACHED = i18n.translate('xpack.cases.caseTable.alreadyAttached', {
+  defaultMessage: 'Added',
```
6be857d961802a2da85815d1057305145459b26f

justinkambic committed:
Introduce io-ts typing for new API response.

justinkambic committed:
Add fpts decoder for new endpoint response.

justinkambic committed:
Add decoder call to API call.

justinkambic commented on x-pack/platform/plugins/shared/cases/public/containers/api.ts:
```diff
@@ -221,6 +222,18 @@ export const getInferenceConnectors = async (
 
+export const findCasesByAttachmentId = async (alertIds: string[], caseIds: string[]) => {
+  return KibanaServices.get().http.fetch<{
+    casesWithAllAttachments: string[];
```
https://github.com/elastic/kibana/pull/231877/commits/4b544a79e72d8e7635e0f4b0657315e9b85ba22e
https://github.com/elastic/kibana/pull/231877/commits/aa237596a4102c6fc090c52715fb68c8e0d38cc7
https://github.com/elastic/kibana/pull/231877/commits/b1e947ed0d3033477c06ceb1b147a2c04b6a76f4

justinkambic commented:
All feedback is implemented. Will solicit review from @elastic/obs-ux-management-team

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/eslint_all_files --no-cache --fix'

dominiqueclarke reviewed (approved):
LGTM

baileycash-elastic reviewed (approved):
lgtm
